File: jashing/src/main/java/com/github/avarabyeu/jashing/utils/InstanceOfMap.java
package com.github.avarabyeu.jashing.utils;

import com.google.common.base.Preconditions;
import com.google.common.collect.ForwardingMap;

import javax.annotation.Nullable;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Contains object as values and it's classes as keys
 * Extends default {@code {java.util.Map<Class<? extends T>, T>}} map with {@link #getInstanceOf(Class)} method
 * which first calls {@link #get(Object)} (and returns value if found) and after goes through entries and
 * tries to find values using {@link java.lang.Class#isAssignableFrom(Class)} construction
 *
 * @author Andrei Varabyeu
 */
public final class InstanceOfMap<T> extends ForwardingMap<Class<? extends T>, T> {
    private Map<Class<? extends T>, T> delegate;

    private InstanceOfMap(Map<Class<? extends T>, T> delegate) {
        this.delegate = delegate;
    }

    @Override
    protected Map<Class<? extends T>, T> delegate() {
        return delegate;
    }

    @Override
    public T put(@Nullable Class<? extends T> clazz, @Nullable T value) {
        return delegate.put(clazz, value);
    }

    public T getInstanceOf(Class<? extends T> clazz) {
        if (delegate.containsKey(clazz)) {
            return delegate.get(clazz);
        } else {
            return delegate.entrySet().stream()
                    .filter(entry -> clazz.isAssignableFrom(entry.getKey()))
                    .map(Map.Entry::getValue)
                    .findFirst().orElse(null);
        }
    }

    public Collection<T> values() {
        return delegate.values();
    }

    public static <T> Builder<T> builder() {
        return new Builder<>();
    }

    public static <T> InstanceOfMap<T> empty() {
        return new InstanceOfMap<>(Collections.emptyMap());
    }

    public static class Builder<T> {

        @SuppressWarnings("unchecked")
        public InstanceOfMap<T> fromList(List<? extends T> list) {
            Preconditions.checkNotNull(list, "Provided list is null!");
            Map<Class<? extends T>, T> map = new HashMap<>(list.size());
            for (T item : list) {
                map.put((Class<? extends T>) item.getClass(), item);
            }
            return new InstanceOfMap<>(map);
        }

        public InstanceOfMap<T> fromArray(T... array) {
            Preconditions.checkNotNull(array, "Provided array is null!");
            return fromList(Arrays.asList(array));
        }

        public InstanceOfMap<T> fromMap(Map<Class<? extends T>, T> map) {
            return new InstanceOfMap<>(map);
        }

    }

}

File: jashing/src/main/kotlin/com/github/avarabyeu/jashing/utils/InstanceOfMap.kt
package com.github.avarabyeu.jashing.utils

import com.google.common.base.Preconditions
import com.google.common.collect.ForwardingMap
import java.util.*

/**
 * Contains object as values and it's classes as keys
 * Extends default `{java.util.Map&lt;Class&lt;? extends T&gt;, T&gt;}` map with [.getInstanceOf] method
 * which first calls [.get] (and returns value if found) and after goes through entries and
 * tries to find values using [java.lang.Class.isAssignableFrom] construction

 * @author Andrei Varabyeu
 */
class InstanceOfMap<T>(val delegate: MutableMap<Class<out T>, T>) : ForwardingMap<Class<out T>, T>() {

    override fun delegate(): Map<Class<out T>, T> {
        return delegate
    }

    override fun put(key: Class<out T>, value: T): T? {
        return delegate.put(key, value)
    }

    fun getInstanceOf(clazz: Class<out T>): T? {
        if (delegate.containsKey(clazz)) {
            return delegate[clazz]
        } else {
            return delegate.entries.filter({ entry -> clazz.isAssignableFrom(entry.key) }).map({ it.value }).firstOrNull();
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other)
            return true
        if (other == null || javaClass != other.javaClass)
            return false
        if (!super.equals(other))
            return false
        val that = other as InstanceOfMap<*>?
        return delegate == that!!.delegate
    }

    override fun hashCode(): Int {
        return Objects.hash(super.hashCode(), delegate)
    }


    class Builder<T> where T : Any {

        fun fromList(list: List<T>): InstanceOfMap<T> {
            Preconditions.checkNotNull(list, "Provided list is null!")
            val map = HashMap<Class<out T>, T>(list.size)
            for (item in list) {

                map.put(item.javaClass as Class<out T>, item)
            }
            return InstanceOfMap(map)
        }

        fun fromArray(array: T): InstanceOfMap<T> {
            Preconditions.checkNotNull(array, "Provided array is null!")
            return fromList(Arrays.asList(array) as List<T>)
        }

        fun fromMap(map: MutableMap<Class<out T>, T>): InstanceOfMap<T> {
            return InstanceOfMap(map)
        }

    }


    companion object {

        @JvmStatic
        fun <T> builder(): Builder<T> where T : Any {
            return Builder()
        }

        @JvmStatic
        fun <T> empty(): InstanceOfMap<T> {
            return InstanceOfMap(mutableMapOf());
        }

    }

}
